apis/python/src/tiledb/vector_search/object_api/object_index.py

```diff
@@ -0,0 +1,390 @@
+
+        if return_objects:
+            unique_objects = self.object_reader.read_objects_by_external_ids(unique_ids)
+            objects = OrderedDict()
+            for attr in unique_objects.keys():
+                objects[attr] = unique_objects[attr][idx]
+
+        if return_objects and return_metadata:
+            return distances, objects, object_metadata
+        elif return_objects and not return_metadata:
+            return distances, objects
+        elif not return_objects and return_metadata:
+            return distances, object_ids, object_metadata
+        elif not return_objects and not return_metadata:
+            return distances, object_ids
+
+    def update_index(
+        self,
+        object_array_timestamp=None,
+        index_timestamp: int = None,
+        workers: int = -1,
+        worker_resources: Dict = None,
+        max_tasks_per_stage: int = -1,
+        verbose: bool = False,
+        trace_id: Optional[str] = None,
+        mode: Mode = Mode.LOCAL,
+        config: Optional[Mapping[str, Any]] = None,
+        namespace: Optional[str] = None,
+        **kwargs,
+    ):
+        embeddings_array_name = storage_formats[self.index.storage_version][
+            "INPUT_VECTORS_ARRAY_NAME"
+        ]
+        embeddings_uri = f"{self.uri}/{embeddings_array_name}"
+        storage_formats[self.index.storage_version]["EXTERNAL_IDS_ARRAY_NAME"]
+        metadata_array_uri = None
+        if self.materialize_object_metadata:
+            metadata_array_uri = self.object_metadata_array_uri
+        if config is None:
+            config = self.config
+        object_api.ingest_embeddings(
+            object_index=self,
+            embeddings_uri=embeddings_uri,
+            metadata_array_uri=metadata_array_uri,
+            index_timestamp=index_timestamp,
+            max_tasks_per_stage=max_tasks_per_stage,
+            workers=workers,
+            worker_resources=worker_resources,
+            verbose=verbose,
+            trace_id=trace_id,
+            mode=mode,
+            config=config,
+            namespace=namespace,
+            **kwargs,
+        )
+        self.index = ingest(
+            index_type=self.index_type,
+            index_uri=self.uri,
+            source_uri=embeddings_uri,
+            source_type="TILEDB_PARTITIONED_ARRAY",
+            external_ids_uri=embeddings_uri,
+            external_ids_type="TILEDB_PARTITIONED_ARRAY",
+            index_timestamp=index_timestamp,
+            # size=size,
+            storage_version=self.index.storage_version,
+            config=config,
+            mode=mode,
+            **kwargs,
+        )
+
+
+def encode_class(a):
+    pickled_object = cloudpickle.dumps(a, protocol=TILEDB_CLOUD_PROTOCOL)
+    return base64.b64encode(pickled_object).decode("ascii")
+
+
+def get_source_code(a):
+    import inspect
+
+    f = open(inspect.getsourcefile(a.__class__), "r")
+    return f.read()
+
+
+def instantiate_object(code, class_name, **kwargs):
```

Comment:
This seems to be the same function as in `ingest_embeddings`?

Reply:
Yes, if we use it as a library the package needs to be released for the library to be found inside cloud UDFs. Until the release goes through I duplicated the code.